File: 351004/Brazhalovich/Lab_1/publisher/src/main/java/org/example/newsapi/service/NewsService.java
package org.example.newsapi.service;

import lombok.RequiredArgsConstructor;
import org.example.newsapi.dto.request.NewsRequestTo;
import org.example.newsapi.dto.response.NewsResponseTo;
import org.example.newsapi.entity.Marker;
import org.example.newsapi.entity.News;
import org.example.newsapi.entity.User;
import org.example.newsapi.exception.AlreadyExistsException;
import org.example.newsapi.exception.NotFoundException;
import org.example.newsapi.mapper.NewsMapper;
import org.example.newsapi.repository.MarkerRepository;
import org.example.newsapi.repository.NewsRepository;
import org.example.newsapi.repository.UserRepository;
import org.springframework.cache.annotation.CacheEvict;
import org.springframework.cache.annotation.Cacheable;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.HashSet;
import java.util.Set;

@Service
@RequiredArgsConstructor
@Transactional(readOnly = true)
public class NewsService {

    private final NewsRepository newsRepository;
    private final UserRepository userRepository;
    private final MarkerRepository markerRepository;
    private final NewsMapper newsMapper;

    @Transactional
    @CacheEvict(value = "news", allEntries = true)
    public NewsResponseTo create(NewsRequestTo request) {
        if (request.getUserId() == null || !userRepository.existsById(request.getUserId())) {
            throw new NotFoundException("User not found");
        }
        if (newsRepository.existsByTitle(request.getTitle())) {
            throw new AlreadyExistsException("News title already exists");
        }

        User user = userRepository.getReferenceById(request.getUserId());
        News news = newsMapper.toEntity(request);
        news.setUser(user);
        news.setCreated(LocalDateTime.now());
        news.setModified(LocalDateTime.now());

        if (request.getMarkerNames() != null && !request.getMarkerNames().isEmpty()) {
            Set<Marker> markers = new HashSet<>();
            for (String name : request.getMarkerNames()) {
                Marker marker = markerRepository.findByName(name)
                        .orElseGet(() -> {
                            Marker newMarker = Marker.builder().name(name).build();
                            return markerRepository.save(newMarker);
                        });
                markers.add(marker);
            }
            news.setMarkers(markers);
        }

        News saved = newsRepository.saveAndFlush(news);
        return newsMapper.toDto(saved);
    }

    public Page<NewsResponseTo> findAll(Pageable pageable) {
        return newsRepository.findAll(pageable).map(newsMapper::toDto);
    }

    @Cacheable(value = "news", key = "#id")
    public NewsResponseTo findById(Long id) {
        return newsRepository.findById(id)
                .map(newsMapper::toDto)
                .orElseThrow(() -> new NotFoundException("News not found"));
    }

    @Transactional
    @CacheEvict(value = "news", allEntries = true)
    public NewsResponseTo update(Long id, NewsRequestTo request) {
        News news = newsRepository.findById(id)
                .orElseThrow(() -> new NotFoundException("News not found"));

        if (!userRepository.existsById(request.getUserId())) {
            throw new NotFoundException("User not found");
        }

        newsMapper.updateEntityFromDto(request, news);
        news.setUser(userRepository.getReferenceById(request.getUserId()));
        news.setModified(LocalDateTime.now());

        if (request.getMarkerNames() != null) {
            Set<Marker> markers = new HashSet<>();
            for (String name : request.getMarkerNames()) {
                Marker marker = markerRepository.findByName(name)
                        .orElseGet(() -> {
                            Marker newMarker = Marker.builder().name(name).build();
                            return markerRepository.save(newMarker);
                        });
                markers.add(marker);
            }
            news.setMarkers(markers);
        }

        return newsMapper.toDto(newsRepository.save(news));
    }

    @Transactional
    @CacheEvict(value = "news", allEntries = true)
    public void delete(Long id) {
        if (!newsRepository.existsById(id)) {
            throw new NotFoundException("News not found");
        }
        newsRepository.deleteById(id);
    }
}

File: 351003/brezgunov/distcomp/src/main/kotlin/com/distcomp/service/NewsService.kt
package com.distcomp.service

import com.distcomp.dto.news.NewsRequestTo
import com.distcomp.dto.news.NewsResponseTo
import com.distcomp.entity.Marker
import com.distcomp.exception.NewsNotFoundException
import com.distcomp.exception.NewsTitleDuplicateException
import com.distcomp.exception.UserNotFoundException
import com.distcomp.exception.ValidationException
import com.distcomp.mapper.NewsMapper
import com.distcomp.repository.NewsRepository
import com.distcomp.repository.UserRepository
import org.springframework.cache.annotation.CacheEvict
import org.springframework.cache.annotation.CachePut
import org.springframework.cache.annotation.Cacheable
import org.springframework.cache.annotation.Caching
import org.springframework.data.repository.findByIdOrNull
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class NewsService(
    val newsMapper: NewsMapper,
    val newsRepository: NewsRepository,
    val userRepository: UserRepository,
) {
    @Transactional
    @CachePut(value = ["news"], key = "#result.id")
    @CacheEvict(value = ["news"], key = "'all'")
    fun createNews(newsRequestTo: NewsRequestTo): NewsResponseTo {
        if (newsRepository.existsByTitle(newsRequestTo.title)) {
            throw NewsTitleDuplicateException("Title already exists")
        }
        val news = newsMapper.toNewsEntity(newsRequestTo)
        val user = userRepository.findByIdOrNull(newsRequestTo.userId)
            ?: throw UserNotFoundException("User not found")
        news.user = user
        if (newsRequestTo.markers != null) {
            for (markerName in newsRequestTo.markers) {
                news.markers.add(Marker(name = markerName))
            }
        }
        newsRepository.save(news)
        return newsMapper.toNewsResponse(news)
    }

    @Cacheable(value = ["news"], key = "#id")
    fun readNewsById(id: Long): NewsResponseTo {
        val news = newsRepository.findByIdOrNull(id) ?: throw NewsNotFoundException("News not found")
        return newsMapper.toNewsResponse(news)
    }

    @Cacheable(value = ["news"], key = "'all'")
    fun readAll(): List<NewsResponseTo> {
        return newsRepository.findAll().map { newsMapper.toNewsResponse(it) }
    }

    @Transactional
    @CachePut(value = ["news"], key = "#newsId")
    @CacheEvict(value = ["news"], key = "'all'")
    fun updateNews(newsRequestTo: NewsRequestTo, newsId: Long?): NewsResponseTo {
        if (newsId == null || newsRepository.findByIdOrNull(newsId) == null) {
            throw NewsNotFoundException("News not found")
        }
        if (newsRequestTo.title.length < 2) {
            throw ValidationException("New title is too short")
        }
        val news = newsMapper.toNewsEntity(newsRequestTo)
        news.id = newsId
        val user = userRepository.findByIdOrNull(newsRequestTo.userId)
        news.user = user
        newsRepository.save(news)
        return newsMapper.toNewsResponse(news)
    }

    @Transactional
    @Caching(evict = [
        CacheEvict(value = ["news"], key = "#id"),
        CacheEvict(value = ["news"], key = "'all'")
    ])
    fun removeNewsById(id: Long) {
        if (newsRepository.findByIdOrNull(id) == null) {
            throw NewsNotFoundException("News not found")
        }
        newsRepository.deleteById(id)
    }
}